core/primitives/src/runtime/parameter_table.rs

```diff
@@ -16,36 +16,90 @@ pub(crate) enum ParameterValue {
+
+    fn try_from(value: &ParameterValue) -> Result<Self, Self::Error> {
+        match value {
+            ParameterValue::U64(v) => Ok(*v),
+            _ => Err(ValueConversionError::WrongType(std::any::type_name::<u64>(), value.clone())),
         }
     }
+}
 
-    fn as_rational(&self) -> Option<Rational32> {
-        match self {
-            ParameterValue::Rational { numerator, denominator } => {
-                Some(Rational32::new(*numerator, *denominator))
-            }
-            _ => None,
+impl TryFrom<&ParameterValue> for u128 {
+    type Error = ValueConversionError;
+
+    fn try_from(value: &ParameterValue) -> Result<Self, Self::Error> {
+        match value {
+            ParameterValue::U64(v) => Ok(u128::from(*v)),
+            ParameterValue::String(s) => s.parse().map_err(|err| {
+                ValueConversionError::ParseInt(err, std::any::type_name::<u128>(), value.clone())
+            }),
+            _ => Err(ValueConversionError::WrongType(std::any::type_name::<u128>(), value.clone())),
         }
     }
+}
 
-    fn as_u128(&self) -> Option<u128> {
-        match self {
-            ParameterValue::U64(v) => Some(u128::from(*v)),
-            // TODO(akashin): Refactor this to use `TryFrom` and properly propagate an error.
-            ParameterValue::String(s) => s.parse().ok(),
-            _ => None,
+// TODO(akashin): Enable this.
+// impl<T> TryFrom<&ParameterValue> for T
+// where
+//     T: TryFrom<u64>,
+//     T::Error: Into<std::num::TryFromIntError>,
+// {
+//     type Error = ValueConversionError;
+//
+//     fn try_from(value: &ParameterValue) -> Result<Self, Self::Error> {
+//         match value {
+//             ParameterValue::U64(v) => T::try_from(*v).map_err(|err| {
+//                 ValueConversionError::TryFromInt(
+//                     err.into(),
+//                     std::any::type_name::<T>(),
+//                     value.clone(),
+//                 )
+//             }),
+//             _ => Err(ValueConversionError::WrongType(std::any::type_name::<T>(), value.clone())),
+//         }
+//     }
+// }
+
+impl TryFrom<&ParameterValue> for Rational32 {
+    type Error = ValueConversionError;
+
+    fn try_from(value: &ParameterValue) -> Result<Self, Self::Error> {
+        match value {
+            ParameterValue::Rational { numerator, denominator } => {
+                Ok(Rational32::new(*numerator, *denominator))
+            }
+            _ => Err(ValueConversionError::WrongType(
+                std::any::type_name::<Rational32>(),
+                value.clone(),
+            )),
         }
     }
+}
+
+impl TryFrom<&ParameterValue> for String {
+    type Error = ValueConversionError;
 
-    fn as_str(&self) -> Option<&str> {
-        match self {
-            ParameterValue::String(v) => Some(v),
-            _ => None,
+    fn try_from(value: &ParameterValue) -> Result<Self, Self::Error> {
+        match value {
+            ParameterValue::String(v) => Ok(v.clone()),
```

Comment:
Since the conversion is taking a `&` in the first place, consider returning a `&str` and give the caller an opportunity to clone if they need to do so. Otherwise convert from `ParameterValue` and return the old value in Err?

Reply:
Thanks for the suggestion! This worked after I sprinkled a few lifetime annotations.